Source/FieldSolver/FiniteDifferenceSolver/EvolveF.cpp

```diff
@@ -0,0 +1,207 @@
+
+        // Extract tileboxes for which to loop
+        Box const& tf  = mfi.tilebox(Ffield->ixType().ixType());
+
+        Real constexpr inv_epsilon0 = 1./PhysConst::ep0;
+
+        // Loop over the cells and update the fields
+        amrex::ParallelFor(tf,
+
+            [=] AMREX_GPU_DEVICE (int i, int j, int k){
+                F(i, j, k) += dt * (
+                    - rho(i, j, k, rhocomp) * inv_epsilon0
+                    + T_Algo::DownwardDx(Ex, coefs_x, n_coefs_x, i, j, k)
+                    + T_Algo::DownwardDy(Ey, coefs_y, n_coefs_y, i, j, k)
+                    + T_Algo::DownwardDz(Ez, coefs_z, n_coefs_z, i, j, k) );
+            }
+
+        );
+
+    }
+
+}
+
+#else // corresponds to ifndef WARPX_DIM_RZ
+
+template<typename T_Algo>
+void FiniteDifferenceSolver::EvolveFCylindrical (
+    std::unique_ptr<amrex::MultiFab>& Ffield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Efield,
+    std::unique_ptr<amrex::MultiFab> const& rhofield,
+    int const rhocomp,
+    amrex::Real const dt ) {
+
+    // Loop through the grids, and over the tiles within each grid
+#ifdef _OPENMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+    for ( MFIter mfi(*Ffield, TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+
+        // Extract field data for this grid/tile
+        Array4<Real> F = Ffield->array(mfi);
+        Array4<Real> const& Er = Efield[0]->array(mfi);
+        Array4<Real> const& Et = Efield[1]->array(mfi);
+        Array4<Real> const& Ez = Efield[2]->array(mfi);
+        Array4<Real> const& rho = rhofield->array(mfi);
+
+        // Extract stencil coefficients
+        Real const * const AMREX_RESTRICT coefs_r = m_stencil_coefs_r.dataPtr();
+        int const n_coefs_r = m_stencil_coefs_r.size();
+        Real const * const AMREX_RESTRICT coefs_z = m_stencil_coefs_z.dataPtr();
+        int const n_coefs_z = m_stencil_coefs_z.size();
+
+        // Extract cylindrical specific parameters
+        Real const dr = m_dr;
+        int const nmodes = m_nmodes;
+        Real const rmin = m_rmin;
+
+        // Extract tileboxes for which to loop
+        Box const& tf  = mfi.tilebox(Ffield->ixType().ixType());
+
+        Real constexpr inv_epsilon0 = 1./PhysConst::ep0;
+        Real constexpr c2 = PhysConst::c * PhysConst::c;
+
+        // Use the right shift in components:
+        // - the first 2*n_rz_azimuthal_modes-1 components correspond to rho old (i.e. rhocomp=0)
+        // - the next 2*n_rz_azimuthal_modes-1 components correspond to rho new (i.e. rhocomp=1)
+        int rho_shift = 0;
+        if (rhocomp == 1) {
+            rho_shift = 2*WarpX::n_rz_azimuthal_modes-1;
+        }
+
+        // Loop over the cells and update the fields
+        amrex::ParallelFor(tf,
+
+            [=] AMREX_GPU_DEVICE (int i, int j, int k){
+                Real const r = rmin + i*dr; // r on a nodal grid (F is nodal in r)
+                if (r != 0) { // Off-axis, regular equations
```

Comment:
Isn't this test a bit dangerous (besides requiring conversion from int to Real), since we don't explicitly set `r` to `0` but actually compute it with operations. We could maybe do something like
```suggestion
                if (abs(r) > dr/100._rt) { // Off-axis, regular equations
```
What do you think?
Note that this also applied to divE.

Reply:
Checking if r != 0 is probably safe since rmin and i will always be >= 0. r will only ever be zero if rmin and i are zero and there shouldn't be any round off issue.